## Devel Branch Differences

This page shows the various branches that have been merged into nh13/Duplex-Sequencing/devel but missing in loeblab/Duplex-Sequencing/master

| Feature | Branch | Pull Request |
| -------- | ------ | ------------ |
| Unpaired reads have their SAM flag set correctly | https://github.com/nh13/Duplex-Sequencing/tree/nh_unpaired_reads | https://github.com/loeblab/Duplex-Sequencing/pull/37 |
| Order of reads at a given site does not affect results | https://github.com/nh13/Duplex-Sequencing/tree/nh_duplex_maker_read_order | https://github.com/loeblab/Duplex-Sequencing/pull/36 |
| Multitude of improvments (see PR) | https://github.com/nh13/Duplex-Sequencing/tree/nh_improvements | https://github.com/loeblab/Duplex-Sequencing/pull/35 |
| Bug Fix | 5eb98e414c3b210d08eb768a3a700b1ad2de8194 | Not submitted |
| Fixups | https://github.com/nh13/Duplex-Sequencing/tree/nh_devel_fixups | Not submitted |
| Bug Fix after removing output BAM in DuplexMaker.py | 8dc7df6689832de882835ec977509cbbf1b7ec35 | Not submitted |
| Add RG to the output of ConsensusMaker if RG exists in the input | 0924cc0987b854c0c35ffbc7becf7c64faec3d61 | Not submitted |
| Make read names consistent across pairs for the SSCS | 81cdc1c3de11f1165d496ecee001b9c1ec8ad8bf | Not submitted |
| Unpaired reads should not have mate information | 6925f9ae159612495f462b2f3c8bf3659389bc08 and 9cebe36031c16cab5f0a7b9ff72d2ddfc1e3ea0b | Not submitted |
